### PR TITLE
Refactor GLUE fine-tune queuing to improve efficiency and add task-specific seed sweeps

### PR DESCRIPTION
**Pull request by alextrott16**
- Runs fine tuning in 2 rounds of parallel jobs (as opposed to `n_pretrain_checkpoints * 2` rounds of parallel jobs). This enables more efficient parallelization, especially when running on machines with `>= 4` GPUs. Concretely, this puts all fine-tuning runs that do not depend on each other in the same round of parallel jobs:
  - The first round of fine-tuning runs train on COLA, MNLI, QNLI, QQP, and SST-2 for all pre-train checkpoints and saves checkpoints for MNLI runs.
  - The second round of fine-tuning trains on RTE, MRPC, and STS-B for all the pre-train checkpoints' downstream MNLI checkpoints (which are only available after the first round).

- Adds `seed_overrides` to the fine-tuning Hparams so the user can specify to sweep over seeds for certain tasks that tend to have noisy outcomes. For example, for each pre-train checkpoint COLA fine-tuning should happen 4 times, each using a different seed.

- This PR includes fixes for a number of "breaking changes" that happened during its development, including issues with multiprocessing, checkpoint naming, and metrics.

### CODE REVIEW

File: examples/glue/run_glue_trainer.py
Comment (moinnadeem) [inline examples/glue/run_glue_trainer.py:224]: Can we move the `zip` and `(ckpt_load_path, parent_ckpt)` into their own line? This line has a lot of work going on.

File: examples/glue/run_glue_trainer.py
Comment (moinnadeem) [inline examples/glue/run_glue_trainer.py:316]: good call on this one

File: composer/yamls/models/glue/stsb.yaml
Comment (moinnadeem) [inline composer/yamls/models/glue/stsb.yaml:55]: Is this only required for STS-B? Or is this an impact of the precision refactor and we should avoid this long-term?

File: examples/glue/run_glue_trainer.py
Comment (moinnadeem) [inline examples/glue/run_glue_trainer.py:239]: should we just deepcopy `train_finetune` and update the seed here, instead of adding an extra parameter to the signature and updating the seed there?

**Review by moinnadeem — APPROVED** [review]
LGTM; thanks for doing this!

**Review by moinnadeem — CHANGES_REQUESTED** [review]
Switching to request changes on the `precision:` stuff since that might be non-trivial for long term

File: composer/yamls/models/glue/stsb.yaml
Comment (alextrott16) [reply composer/yamls/models/glue/stsb.yaml:55]: It seems to be a consequence of the precision refactor (STS-B wasn't causing problems before that), but the issue remains after merging in dev after a fix was pushed. It's unclear to me what the root of the issue is, but it has something to do with the `SpearmanCorrCoeff` metric wanting the model predictions to be `fp32`. I'm not sure how to prevent this issue at the code level, but using `fp32` prevents the error and adds only a small overhead for this quick fine-tuning task.

Given the minor impact, I was hoping we could leave this as a TODO. If it needs to be dealt with now, I'm open to suggestions on how to proceed :)

File: examples/glue/run_glue_trainer.py
Comment (alextrott16) [reply examples/glue/run_glue_trainer.py:239]: Given the way the `merge_hparams` function works, I'm not sure it would work to do the deepcopy strategy. My thinking was also that, because the downstream `train_finetune` function already handles several overwrites, it would be cleaner to handle the seed overwrite in that area of the code.

File: composer/yamls/models/glue/stsb.yaml
Comment (moinnadeem) [reply composer/yamls/models/glue/stsb.yaml:55]: What concerns me isn't the small change on this PR, but rather, it indicates that something about the root cause of the issue still isn't fixed, and it'll be way worse if an open-source user is running into this -- it's highly less likely that they won't know what to do! In other words, we can continue being productive but Composer overall will suffer.

Since we can run our weekend runs without this, @mvpatel2000 thoughts on working with Alex on root-causing and finding an appropriate Composer solution that doesn't require `precision: fp32` here?

File: examples/glue/run_glue_trainer.py
Comment (moinnadeem) [reply examples/glue/run_glue_trainer.py:239]: That is fair, I can go ahead and resolve this one!

File: examples/glue/run_glue_trainer.py
Comment (ishanashastri) [inline examples/glue/run_glue_trainer.py:220]: Curious why we need this wrapper for the lambda function?

Comment (ishanashastri) [inline examples/glue/run_glue_trainer.py:220]: Curious why we need this wrapper for the lambda?

File: composer/yamls/models/glue/stsb.yaml
Comment (alextrott16) [reply composer/yamls/models/glue/stsb.yaml:55]: Very good points!

File: examples/glue/run_glue_trainer.py
Comment (ishanashastri) [inline examples/glue/run_glue_trainer.py:230]: Would it make sense to instead put all the tasks in `seed_overrides,` even if it only needs a single seed, and then remove this extra if-else loop? This also may make yaml vs entrypoint details less confusing. See mostly duplicate code so just thinking about how to collapse.

File: examples/glue/run_glue_trainer.py
Comment (ishanashastri) [inline examples/glue/run_glue_trainer.py:305]: Do we also need to save the seed information here and in wandb?

File: examples/glue/run_glue_trainer.py
Comment (ishanashastri) [inline examples/glue/run_glue_trainer.py:203]: Wondering if it makes sense to make a field in `finetune_hparams` to store these overrides along with all the other overrides? That way, someone is not bound to spawning all these seeds if they don't need to run all the different seedings.

**Review by ishanashastri — COMMENTED** [review]
Overall looks really good, thanks for putting this together so quickly! Mostly comments about seeds and where those overrides should lie.

File: examples/glue/run_glue_trainer.py
Comment (ravi-mosaicml) [reply examples/glue/run_glue_trainer.py:220]: For readability, I think it would be helpful to hardcode the lambda directly in twice, even though it would be repeated code.

**Review by ravi-mosaicml — APPROVED** [review]
Overall looks fine; I trust that this works, but see comments re: readability.
The inline comments are very helpful.

File: examples/glue/run_glue_trainer.py
Comment (alextrott16) [reply examples/glue/run_glue_trainer.py]: Added this change

File: examples/glue/run_glue_trainer.py
Comment (alextrott16) [reply examples/glue/run_glue_trainer.py]: Removed the wrapper

File: examples/glue/run_glue_trainer.py
Comment (ishanashastri) [reply examples/glue/run_glue_trainer.py:204]: Will sync offline about this!

**Review by moinnadeem — APPROVED** [review]
LGTM, let's move the precision context into a new issue so we can unblock. We do need to address the precision for sure

**Review by hanlint — APPROVED** [review]

**Pull request merged**
